src/winhello/WindowsHello.cpp

```diff
@@ -0,0 +1,202 @@
+    {
+        error.clear();
+        auto challengeBuffer = CryptographicBuffer::CreateFromByteArray(
+            array_view<uint8_t>(reinterpret_cast<uint8_t*>(challenge.data()), challenge.size()));
+
+        return AsyncTask::runAndWaitForFuture([&] {
+            // The first time this is used a key-pair will be generated using the common name
+            auto result =
+                KeyCredentialManager::RequestCreateAsync(s_winHelloKeyName, KeyCredentialCreationOption::FailIfExists)
+                    .get();
+
+            if (result.Status() == KeyCredentialStatus::CredentialAlreadyExists) {
+                result = KeyCredentialManager::OpenAsync(s_winHelloKeyName).get();
+            } else if (result.Status() != KeyCredentialStatus::Success) {
+                error = QObject::tr("Failed to create Windows Hello credential.");
+                return false;
+            }
+
+            const auto signature = result.Credential().RequestSignAsync(challengeBuffer).get();
+            if (signature.Status() != KeyCredentialStatus::Success) {
+                error = QObject::tr("Failed to sign challenge using Windows Hello.");
+                return false;
+            }
+
+            // Use the SHA-256 hash of the challenge signature as the encryption key
+            const auto response = signature.Result();
+            CryptoHash hasher(CryptoHash::Sha256);
+            hasher.addData({reinterpret_cast<const char*>(response.data()), static_cast<int>(response.Length())});
+            key = hasher.result();
+            return true;
+        });
+    }
+} // namespace
+
+WindowsHello* WindowsHello::m_instance{nullptr};
+WindowsHello* WindowsHello::instance()
+{
+    if (!m_instance) {
+        m_instance = new WindowsHello();
+    }
+    return m_instance;
+}
+
+WindowsHello::WindowsHello(QObject* parent)
+    : QObject(parent)
+{
+    concurrency::create_task([this] {
+        bool state = KeyCredentialManager::IsSupportedAsync().get();
+        m_available = state;
+        emit availableChanged(m_available);
+    });
+}
+
+bool WindowsHello::isAvailable() const
+{
+    return m_available;
+}
+
+QString WindowsHello::errorString() const
+{
+    return m_error;
+}
+
+bool WindowsHello::storeKey(const QString& dbPath, const QByteArray& data)
+{
+    queueSecurityPromptFocus();
+
+    // Generate a random challenge that will be signed by Windows Hello
+    // to create the key. The challenge is also used as the IV.
+    auto ivSize = SymmetricCipher::defaultIvSize(SymmetricCipher::Aes256_CBC);
+    auto challenge = Random::instance()->randomArray(ivSize);
+    QByteArray key;
+    if (!deriveEncryptionKey(challenge, key, m_error)) {
+        return false;
+    }
+
+    // Encrypt the data using AES-256-CBC
+    SymmetricCipher cipher;
+    if (!cipher.init(SymmetricCipher::Aes256_CBC, SymmetricCipher::Encrypt, key, challenge)) {
+        m_error = tr("Failed to init KeePassXC crypto.");
+        return false;
+    }
```

Comment:
I think this should be GCM, not CBC, otherwise you have to add a MAC.

Reply:
Great point, added the AES-256/GCM algorithm to SymmetricCipher and used it here and in TouchID.